python/benchmark/gen_data_distributed.py

```diff
@@ -379,33 +392,77 @@ def gen_dataframe_and_meta(
 
         if shuffle:
             # Shuffle feature indices upfront.
-            indices = np.arange(cols)
-            generator.shuffle(indices)
-            ground_truth = ground_truth[indices]
+            col_indices = np.arange(cols)
+            generator.shuffle(col_indices)
+            ground_truth = ground_truth[col_indices]
+
+        # Create different partition seeds for sample generation.
+        random.seed(params["random_state"])
+        seed_maxval = 100 * num_partitions
+        partition_seeds = random.sample(range(1, seed_maxval), num_partitions)
 
         # UDF for distributed generation of X and y.
         def make_regression_udf(iter: Iterable[pd.DataFrame]) -> Iterable[pd.DataFrame]:
             for pdf in iter:
-                X_p = pdf.to_numpy()
+                use_cupy = use_gpu
+                if use_cupy:
+                    try:
+                        import cupy as cp
+                    except ImportError:
+                        use_cupy = False
+                        logging.warning("cupy import failed; falling back to numpy.")
+
+                partition_index = pyspark.TaskContext().partitionId()
+
+                if use_cupy:
+                    X_p = cp.asarray(pdf.to_numpy())
+                    generator = cp.random.RandomState(partition_seeds[partition_index])
+                else:
+                    X_p = pdf.to_numpy()
+                    generator = np.random.RandomState(partition_seeds[partition_index])
+
                 n_partition_rows = X_p.shape[0]
 
                 if shuffle:
                     # Column-wise shuffle (global)
-                    X_p[:, :] = X_p[:, indices]
+                    if use_cupy:
+                        X_p[:, :] = X_p[:, cp.asarray(col_indices)]
+                    else:
+                        X_p[:, :] = X_p[:, col_indices]
 
-                y = np.dot(X_p, ground_truth) + bias
+                if use_cupy:
+                    y = cp.dot(X_p, cp.asarray(ground_truth)) + bias
```

Comment:
Do you mean outside the `for pdf in iter` loop? Might be misunderstanding, but since this would move the operation it outside of the partition, would it execute on the driver side, which we want to avoid w/cupy operations?

Reply:
Everything in the udf runs in the executor.  The loop is over the arrow batching within the partition.